FWCore/Framework/interface/ESProducer.h

```diff
@@ -193,16 +233,32 @@ namespace edm {
         method in order to do the registration with the EventSetup
     */
     template <typename T, typename TReturn, typename TRecord, typename TArg>
-    ESConsumesCollectorT<TRecord> setWhatProduced(T* iThis,
-                                                  TReturn (T ::*iMethod)(const TRecord&),
-                                                  const TArg& iDec,
-                                                  const es::Label& iLabel = {}) {
+    auto setWhatProduced(T* iThis,
+                         TReturn (T ::*iMethod)(const TRecord&),
+                         const TArg& iDec,
+                         const es::Label& iLabel = {}) {
+      return setWhatProduced<TReturn, TRecord>(
+          [iThis, iMethod](TRecord const& iRecord) { return (iThis->*iMethod)(iRecord); },
+          createDecoratorFrom(iThis, static_cast<const TRecord*>(nullptr), iDec),
+          iLabel);
+    }
+
+    template <typename TFunc>
+    auto setWhatProduced(TFunc&& func, const es::Label& iLabel = {}) {
+      using Types = eventsetup::detail::RetArgType<TFunc>;
+      using TReturn = typename Types::return_type;
+      using TRecord = typename Types::argument_type;
+      using DecoratorType = eventsetup::CallbackSimpleDecorator<TRecord>;
+      return setWhatProduced<TReturn, TRecord>(std::forward<TFunc>(func), DecoratorType(), iLabel);
+    }
+
+    template <typename TReturn, typename TRecord, typename TFunc, typename TDecorator>
+    ESConsumesCollectorT<TRecord> setWhatProduced(TFunc&& func, TDecorator&& iDec, const es::Label& iLabel = {}) {
       const auto id = consumesInfos_.size();
-      using DecoratorType = typename eventsetup::DecoratorFromArg<T, TRecord, TArg>::Decorator_t;
-      using CallbackType = eventsetup::Callback<T, TReturn, TRecord, DecoratorType>;
+      using DecoratorType = std::decay_t<TDecorator>;
+      using CallbackType = eventsetup::Callback<ESProducer, TFunc, TReturn, TRecord, DecoratorType>;
```

Comment:
Using `ESProducer` rather than `T` probably doesn't really save any space and requires the compiler to call virtual functions through the virtual function table rather than directly (if they are marked `final`).

Reply:
The change certainly does not save space since `Callback` stores a pointer to the producer. But all functions (other than the produce methdo call) called by `Callback` are of `ESProducer`and non-virtual, so there should be no impact on function call performance.

I'm more annoyed by the pointer-to-producer being stored twice now, once as a member of the `Callback` and a second time in the capture of the lambda. In principle it should be possible to add an "member function or a standalone functor" abstraction that would preserve the current structure of `Callback` for producer member functions, but I'm not sure if that would be worth it.

Reply:
I opened an issue #39221 based on our discussion yesterday on the first template argument of `Callback`.